source/common/upstream/health_checker_base_impl.cc

```diff
@@ -55,6 +56,20 @@ HealthCheckerImplBase::initTransportSocketOptions(
   return std::make_shared<const Network::TransportSocketOptionsImpl>();
 }
 
+absl::optional<MetadataConstSharedPtr> HealthCheckerImplBase::initTransportSocketMatchMetadata(
+    const envoy::config::core::v3::HealthCheck& config) {
+  if (config.has_transport_socket_match_criteria()) {
+    std::unique_ptr<envoy::config::core::v3::Metadata> metadata =
```

Comment:
Whether `metadata` is a shared ptr or unique ptr, I still need to end with a move because I need to mutate the underlying object but ultimately want to return a pointer to a const. 

Since the move is required either way, making `metadata` a unique ptr made the most sense to me because it doesn't need to be shared.

What do you think? Does that line of thinking make sense or am I missing something?

Reply:
You should be able to do that without a move, e.g. https://godbolt.org/z/uu8qJ5

Reply:
Oh, neat.